# Topic Modeling

## Introduction

Another popular text analysis technique is called topic modeling. The ultimate goal of topic modeling is to find various topics that are present in your corpus. Each document in the corpus will be made up of at least one topic, if not multiple topics.

In this notebook, we will be covering the steps on how to do **Latent Dirichlet Allocation (LDA)**, which is one of many topic modeling techniques. It was specifically designed for text data.

To use a topic modeling technique, you need to provide (1) a document-term matrix and (2) the number of topics you would like the algorithm to pick up.

Once the topic modeling technique is applied, your job as a human is to interpret the results and see if the mix of words in each topic make sense. If they don't make sense, you can try changing up the number of topics, the terms in the document-term matrix, model parameters, or even try a different model.

## Topic Modeling - Attempt #1 (All Text)

In [83]:
# 1.Let's read in our document-term matrix
import pandas as pd
import pickle

data = pd.read_pickle('dtm_stop.pkl')
data

abandoning  abe  abhorred  ability  abitur  able  ably  abolish  \
abraham              0    1         1        1       0     1     0        0   
ambedkar             0    0         0        0       0     0     1        0   
boris                0    0         0        0       0     0     0        0   
brandt               0    0         0        0       1     2     0        0   
desmond              0    0         0        0       0     1     0        0   
gandhi               0    0         0        0       0     0     0        0   
mandela              0    0         0        1       0     1     0        0   
margaret             1    0         0        0       0     0     0        0   
roosevelt            0    0         0        0       0     2     0        1   
trump                0    0         0        0       0     1     0        0   
washington           0    0         0        0       0     2     0        0   
winston              0    0         0        0       0     1     0        0   

            abolitionists  abomination  ...  young  youngster  youre  \
abraham                 1            0  ...      1          0      0   
ambedkar                0            0  ...      2          0      0   
boris                   0            0  ...      1          0      0   
brandt                  0            0  ...      2          0      0   
desmond                 0            1  ...      0          0      0   
gandhi                  0            0  ...      1          1      0   
mandela                 0            0  ...      1          1      0   
margaret                0            0  ...      0          0      0   
roosevelt               0            0  ...      0          0      0   
trump                   0            0  ...      0          0      1   
washington              0            0  ...      1          0      0   
winston                 0            0  ...      1          0      0   

            yousafzai  youth  zeal  zealand  zelnickova  zone  zulu  
abraham             0      0     0        0           0     0     0  
ambedkar            0      0     0        0           0     0     0  
boris               0      0     0        0           0     0     0  
brandt              0      1     1        0           0     0     0  
desmond             0      0     0        0           0     0     0  
gandhi              0      0     0        0           0     0     1  
mandela             1      0     0        1           0     0     0  
margaret            0      0     0        0           0     1     0  
roosevelt           0      0     0        0           0     0     0  
trump               0      0     0        0           1     0     0  
washington          0      0     0        0           0     0     0  
winston             0      0     0        0           0     0     0  

[12 rows x 3650 columns]

In [84]:
#2
# Import the necessary modules for LDA with gensim
# Terminal / Anaconda Navigator: conda install -c conda-forge gensim
# matutils --> Converting Data Structures To and From NumPy Arrays
# scipy.sparse --> for handling Sparse Matrics
from gensim import matutils, models
import scipy.sparse

# import logging
# logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)

In [85]:
#3
# One of the required inputs is a term-document matrix
tdm = data.transpose()
tdm.head()

abraham  ambedkar  boris  brandt  desmond  gandhi  mandela  \
abandoning        0         0      0       0        0       0        0   
abe               1         0      0       0        0       0        0   
abhorred          1         0      0       0        0       0        0   
ability           1         0      0       0        0       0        1   
abitur            0         0      0       1        0       0        0   

            margaret  roosevelt  trump  washington  winston  
abandoning         1          0      0           0        0  
abe                0          0      0           0        0  
abhorred           0          0      0           0        0  
ability            0          0      0           0        0  
abitur             0          0      0           0        0

In [86]:
#4
# We're going to put the term-document matrix into a new gensim format, from df --> sparse matrix --> gensim corpus
sparse_counts = scipy.sparse.csr_matrix(tdm)
corpus = matutils.Sparse2Corpus(sparse_counts)

In [87]:
#5
# Gensim also requires dictionary of the all terms and their respective location in the term-document matrix
cv = pickle.load(open("cv_stop.pkl", "rb"))
id2word = dict((v, k) for k, v in cv.vocabulary_.items())

Now that we have the corpus (term-document matrix) and id2word (dictionary of location: term), we need to specify two other parameters - the number of topics and the number of passes. Let's start the number of topics at 2, see if the results make sense, and increase the number from there.

In [88]:
#6
# Now that we have the corpus (term-document matrix) and id2word (dictionary of location: term),
# we need to specify two other parameters as well - the number of topics and the number of passes
lda = models.LdaModel(corpus=corpus, id2word=id2word, num_topics=2, passes=10)
lda.print_topics()

[(0,
  '0.009*"churchill" + 0.008*"lincoln" + 0.008*"war" + 0.008*"mandela" + 0.008*"gandhi" + 0.006*"washington" + 0.005*"nelson" + 0.005*"south" + 0.005*"world" + 0.004*"roosevelt"'),
 (1,
  '0.007*"brandt" + 0.007*"johnson" + 0.007*"trump" + 0.006*"boris" + 0.006*"tutu" + 0.005*"thatcher" + 0.004*"party" + 0.004*"ambedkar" + 0.004*"german" + 0.003*"london"')]

In [89]:
#7
# LDA for num_topics = 3
lda = models.LdaModel(corpus=corpus, id2word=id2word, num_topics=3, passes=10)
lda.print_topics()

[(0,
  '0.009*"war" + 0.008*"churchill" + 0.008*"gandhi" + 0.008*"lincoln" + 0.007*"brandt" + 0.005*"washington" + 0.005*"world" + 0.004*"indian" + 0.004*"independence" + 0.004*"india"'),
 (1,
  '0.014*"mandela" + 0.013*"johnson" + 0.012*"boris" + 0.009*"nelson" + 0.009*"thatcher" + 0.006*"south" + 0.005*"party" + 0.005*"london" + 0.005*"mrs" + 0.005*"africa"'),
 (2,
  '0.017*"trump" + 0.015*"tutu" + 0.008*"desmond" + 0.006*"south" + 0.005*"donald" + 0.003*"africa" + 0.003*"black" + 0.003*"business" + 0.003*"presidential" + 0.003*"campaign"')]

In [90]:
#8
# LDA for num_topics = 4
lda = models.LdaModel(corpus=corpus, id2word=id2word, num_topics=4, passes=10)
lda.print_topics()

[(0,
  '0.021*"lincoln" + 0.015*"washington" + 0.007*"slavery" + 0.005*"war" + 0.005*"george" + 0.005*"president" + 0.005*"states" + 0.004*"abraham" + 0.004*"union" + 0.004*"new"'),
 (1,
  '0.013*"churchill" + 0.011*"gandhi" + 0.011*"johnson" + 0.010*"boris" + 0.009*"tutu" + 0.008*"war" + 0.006*"south" + 0.005*"london" + 0.005*"independence" + 0.005*"desmond"'),
 (2,
  '0.017*"mandela" + 0.012*"nelson" + 0.011*"thatcher" + 0.008*"roosevelt" + 0.006*"south" + 0.006*"world" + 0.006*"mrs" + 0.006*"africa" + 0.005*"president" + 0.005*"time"'),
 (3,
  '0.014*"brandt" + 0.013*"trump" + 0.007*"german" + 0.007*"ambedkar" + 0.005*"germany" + 0.005*"caste" + 0.004*"indian" + 0.004*"willy" + 0.004*"india" + 0.004*"donald"')]

These topics aren't looking too great. We've tried modifying our parameters. Let's try modifying our terms list as well.

## Topic Modeling - Attempt #2 (Nouns Only)

One popular trick is to look only at terms that are from one part of speech (only nouns, only adjectives, etc.). Check out the UPenn tag set: https://www.ling.upenn.edu/courses/Fall_2003/ling001/penn_treebank_pos.html.

In [91]:
#9
# Let's create a function to pull out nouns from a string of text
# NN	Noun, singular or mass   	-JJ	Adjective  	-VB	Verb, base form    	-RB	Adverb
from nltk import word_tokenize, pos_tag

def nouns(text):
    '''Given a string of text, tokenize the text and pull out only the nouns.'''
    is_noun = lambda pos: pos[:2] == 'NN'
    tokenized = word_tokenize(text)
    all_nouns = [word for (word, pos) in pos_tag(tokenized) if is_noun(pos)]
    return ' '.join(all_nouns)

In [92]:
#10
# Read in the cleaned data, before the CountVectorizer step
data_clean = pd.read_pickle('data_clean.pkl')
data_clean

transcript
abraham     with malice toward none with charity for all w...
ambedkar    dr br ambedkar  –  br ambedkar babasaheb was a...
boris       boris johnson is a leading conservative politi...
brandt      willy brandt  – german statesman and politicia...
desmond     desmond mpilo tutu  –  was born in klerksdorp ...
gandhi      mahatma gandhi was a prominent indian politica...
mandela      nelson mandela  –  was a south african politi...
margaret    margaret thatcher  was britains first female p...
roosevelt   franklin delano roosevelt january   – april   ...
trump       donald trump  –  is the  president of the us f...
washington  george washington – was commander in chief of ...
winston     sir winston  churchill  november  –  january  ...

In [93]:
#11
# downloading the necessary resources for NLTK.
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('stopwords')


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\pulki\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\pulki\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\pulki\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [94]:
#12
# Apply the nouns function to the transcripts to filter only on nouns
data_nouns = pd.DataFrame(data_clean.transcript.apply(nouns))
data_nouns

transcript
abraham     none charity firmness right god let work natio...
ambedkar    dr br ambedkar – br ambedkar babasaheb reforme...
boris       boris johnson politician minister leader party...
brandt      statesman politician brandt leftwing politicia...
desmond     desmond mpilo tutu – klerksdorp transvaal octo...
gandhi      mahatma gandhi leader figure campaign independ...
mandela     mandela – activist years prison opposition apa...
margaret    margaret thatcher britains minister views iron...
roosevelt   delano – april initials fdr thirtysecond presi...
trump       trump – president years chairman president tru...
washington  george washington – commander chief forces war...
winston     winston churchill november – january politicia...

In [95]:
#13
# Define custom stop words relevant to biographies of politicians
custom_stop_words = {
    'politician', 'politicians', 'biography', 'biographies',  # Common terms related to the dataset
    'political', 'government', 'president', 'minister', 'senator', 'congressman',  # Political titles
}


# Optionally, you can also add common English stopwords
from nltk.corpus import stopwords
english_stop_words = set(stopwords.words('english'))

# Combine custom and English stopwords
stop_words = custom_stop_words.union(english_stop_words)

# Convert stop_words to a list
stop_words_list = list(stop_words)


In [96]:
#14
# Create a new document-term matrix using only nouns
from sklearn.feature_extraction.text import CountVectorizer

# Convert stop_words to a list
stop_words_list = list(stop_words)

# Recreate a document-term matrix with only nouns
cvn = CountVectorizer(stop_words=stop_words_list)
data_cvn = cvn.fit_transform(data_nouns.transcript)
data_dtmn = pd.DataFrame(data_cvn.toarray(), columns=cvn.get_feature_names_out())
data_dtmn.index = data_nouns.index
data_dtmn


abe  ability  abolitionists  abomination  abraham  absence  abuse  \
abraham       1        1              1            0        5        0      0   
ambedkar      0        0              0            0        0        0      0   
boris         0        0              0            0        0        0      0   
brandt        0        0              0            0        0        0      0   
desmond       0        0              0            1        0        0      1   
gandhi        0        0              0            0        0        0      0   
mandela       0        1              0            0        0        1      0   
margaret      0        0              0            0        0        0      0   
roosevelt     0        0              0            0        1        0      0   
trump         0        0              0            0        1        0      0   
washington    0        0              0            0        0        0      0   
winston       0        0              0            0        0        0      0   

            acceptance  accommodation  accord  ...  yorker  youngster  youre  \
abraham              0              0       0  ...       0          0      0   
ambedkar             0              1       0  ...       0          0      0   
boris                0              0       0  ...       0          0      0   
brandt               1              0       0  ...       0          0      0   
desmond              0              0       0  ...       1          0      0   
gandhi               0              0       0  ...       0          1      0   
mandela              0              0       0  ...       0          1      0   
margaret             0              0       0  ...       0          0      0   
roosevelt            0              0       0  ...       0          0      0   
trump                0              0       1  ...       0          0      1   
washington           0              0       0  ...       0          0      0   
winston              0              0       0  ...       0          0      0   

            yousafzai  youth  zeal  zealand  zelnickova  zone  zulu  
abraham             0      0     0        0           0     0     0  
ambedkar            0      0     0        0           0     0     0  
boris               0      0     0        0           0     0     0  
brandt              0      1     1        0           0     0     0  
desmond             0      0     0        0           0     0     0  
gandhi              0      0     0        0           0     0     1  
mandela             1      0     0        1           0     0     0  
margaret            0      0     0        0           0     1     0  
roosevelt           0      0     0        0           0     0     0  
trump               0      0     0        0           1     0     0  
washington          0      0     0        0           0     0     0  
winston             0      0     0        0           0     0     0  

[12 rows x 2082 columns]

In [97]:
#15
# Create the gensim corpus
corpusn = matutils.Sparse2Corpus(scipy.sparse.csr_matrix(data_dtmn.transpose()))

# Create the vocabulary dictionary
id2wordn = dict((v, k) for k, v in cvn.vocabulary_.items())

In [98]:
#16
# Let's start with 2 topics
ldan = models.LdaModel(corpus=corpusn, num_topics=2, id2word=id2wordn, passes=10)
ldan.print_topics()

[(0,
  '0.014*"mandela" + 0.014*"johnson" + 0.013*"trump" + 0.010*"tutu" + 0.009*"people" + 0.009*"boris" + 0.007*"thatcher" + 0.006*"party" + 0.006*"nelson" + 0.006*"time"'),
 (1,
  '0.015*"war" + 0.011*"churchill" + 0.011*"gandhi" + 0.010*"lincoln" + 0.008*"brandt" + 0.008*"washington" + 0.008*"world" + 0.008*"people" + 0.007*"independence" + 0.005*"roosevelt"')]

In [99]:
#17
# Let's try topics = 3
ldan = models.LdaModel(corpus=corpusn, num_topics=3, id2word=id2wordn, passes=10)
ldan.print_topics()

[(0,
  '0.016*"lincoln" + 0.014*"trump" + 0.012*"brandt" + 0.011*"war" + 0.008*"world" + 0.008*"people" + 0.008*"thatcher" + 0.007*"party" + 0.007*"roosevelt" + 0.006*"germany"'),
 (1,
  '0.016*"mandela" + 0.016*"gandhi" + 0.015*"churchill" + 0.012*"war" + 0.011*"washington" + 0.008*"independence" + 0.008*"people" + 0.007*"world" + 0.006*"nelson" + 0.006*"india"'),
 (2,
  '0.027*"johnson" + 0.019*"tutu" + 0.017*"boris" + 0.010*"desmond" + 0.010*"people" + 0.010*"london" + 0.007*"party" + 0.006*"time" + 0.005*"oxford" + 0.005*"vote"')]

In [100]:
#18
# Let's try 4 topics
ldan = models.LdaModel(corpus=corpusn, num_topics=4, id2word=id2wordn, passes=10)
ldan.print_topics()

[(0,
  '0.022*"johnson" + 0.021*"trump" + 0.014*"boris" + 0.008*"caste" + 0.008*"london" + 0.007*"ambedkar" + 0.007*"party" + 0.006*"people" + 0.005*"campaign" + 0.005*"untouchables"'),
 (1,
  '0.024*"lincoln" + 0.020*"washington" + 0.014*"war" + 0.011*"roosevelt" + 0.011*"people" + 0.009*"slavery" + 0.009*"world" + 0.008*"george" + 0.007*"union" + 0.007*"states"'),
 (2,
  '0.029*"gandhi" + 0.023*"tutu" + 0.012*"desmond" + 0.012*"independence" + 0.010*"people" + 0.007*"india" + 0.006*"africa" + 0.006*"war" + 0.005*"time" + 0.005*"apartheid"'),
 (3,
  '0.019*"mandela" + 0.018*"churchill" + 0.017*"war" + 0.015*"brandt" + 0.010*"world" + 0.009*"thatcher" + 0.009*"party" + 0.008*"nelson" + 0.008*"people" + 0.008*"time"')]

In [101]:
# Let's take a look at which topics each transcript contains
corpus_transformed = ldan[corpusn]
topics_and_indices = []
for doc_topics in corpus_transformed:
    topic_index, _ = max(doc_topics, key=lambda x: x[1])
    topics_and_indices.append(topic_index)

# Now, zip the topics with the document indices
result = list(zip(topics_and_indices, data_dtmn.index))

# Print the result
print(result)

[(1, 'abraham'), (0, 'ambedkar'), (0, 'boris'), (3, 'brandt'), (2, 'desmond'), (2, 'gandhi'), (3, 'mandela'), (3, 'margaret'), (1, 'roosevelt'), (0, 'trump'), (1, 'washington'), (3, 'winston')]


* Topic 0: Focuses on contemporary political figures like Boris Johnson and Donald Trump, as well as discussions related to political parties and caste issues.
* Topic 1: Highlights historical figures such as Abraham Lincoln and George Washington, along with references to slavery, wartime, and the Union.
* Topic 2: Delves into social justice movements, featuring figures like Mahatma Gandhi and Desmond Tutu, and themes related to independence struggles and apartheid.
* Topic 3: Encompasses global events and leadership, including Nelson Mandela, Winston Churchill, and Willy Brandt, suggesting discussions around wartime, political leadership, and historical contexts.

## Topic Modeling - Attempt #3 (Nouns and Adjectives)

In [102]:
#19
# Let's create a function to pull out nouns and Adjective from a string of text
def nouns_adj(text):
    '''Given a string of text, tokenize the text and pull out only the nouns and adjectives.'''
    is_noun_adj = lambda pos: pos[:2] == 'NN' or pos[:2] == 'JJ'
    tokenized = word_tokenize(text)
    nouns_adj = [word for (word, pos) in pos_tag(tokenized) if is_noun_adj(pos)]
    return ' '.join(nouns_adj)

In [103]:
#20
# Apply the nouns function to the transcripts to filter only on nouns
data_nouns_adj = pd.DataFrame(data_clean.transcript.apply(nouns_adj))
data_nouns_adj

transcript
abraham     malice none charity firmness right god right l...
ambedkar    dr br ambedkar – br ambedkar babasaheb indian ...
boris       boris johnson conservative politician british ...
brandt      brandt – german statesman politician brandt ge...
desmond     desmond mpilo tutu – klerksdorp transvaal octo...
gandhi      mahatma gandhi prominent indian political lead...
mandela     nelson mandela – south african political activ...
margaret    margaret thatcher britains first female prime ...
roosevelt   franklin delano january – april initials fdr t...
trump       donald trump – president many years chairman p...
washington  george washington – commander chief continenta...
winston     sir winston churchill november – january briti...

In [104]:
#21
# Create a new document-term matrix using only nouns and adjectives, also remove common words with max_df
cvna = CountVectorizer(stop_words=stop_words_list, max_df=.8)
data_cvna = cvna.fit_transform(data_nouns_adj.transcript)
data_dtmna = pd.DataFrame(data_cvna.toarray(), columns=cvna.get_feature_names_out())
data_dtmna.index = data_nouns_adj.index
data_dtmna


abe  ability  able  abolitionists  abomination  abraham  absence  \
abraham       1        1     1              1            0        6        0   
ambedkar      0        0     0              0            0        0        0   
boris         0        0     0              0            0        0        0   
brandt        0        0     2              0            0        0        0   
desmond       0        0     1              0            1        0        0   
gandhi        0        0     0              0            0        0        0   
mandela       0        1     1              0            0        0        1   
margaret      0        0     0              0            0        0        0   
roosevelt     0        0     2              0            0        1        0   
trump         0        0     1              0            0        1        0   
washington    0        0     2              0            0        0        0   
winston       0        0     1              0            0        0        0   

            abundant  abuse  acceptance  ...  young  youngster  youre  \
abraham            0      0           0  ...      1          0      0   
ambedkar           0      0           0  ...      2          0      0   
boris              0      0           0  ...      1          0      0   
brandt             0      0           1  ...      2          0      0   
desmond            0      1           0  ...      0          0      0   
gandhi             0      0           0  ...      1          1      0   
mandela            0      0           0  ...      1          1      0   
margaret           0      0           0  ...      0          0      0   
roosevelt          0      0           0  ...      0          0      0   
trump              0      0           0  ...      0          0      1   
washington         1      0           0  ...      1          0      0   
winston            0      0           0  ...      1          0      0   

            yousafzai  youth  zeal  zealand  zelnickova  zone  zulu  
abraham             0      0     0        0           0     0     0  
ambedkar            0      0     0        0           0     0     0  
boris               0      0     0        0           0     0     0  
brandt              0      1     1        0           0     0     0  
desmond             0      0     0        0           0     0     0  
gandhi              0      0     0        0           0     0     1  
mandela             1      0     0        1           0     0     0  
margaret            0      0     0        0           0     1     0  
roosevelt           0      0     0        0           0     0     0  
trump               0      0     0        0           1     0     0  
washington          0      0     0        0           0     0     0  
winston             0      0     0        0           0     0     0  

[12 rows x 2684 columns]

In [105]:
#22
# Create the gensim corpus
corpusna = matutils.Sparse2Corpus(scipy.sparse.csr_matrix(data_dtmna.transpose()))

# Create the vocabulary dictionary
id2wordna = dict((v, k) for k, v in cvna.vocabulary_.items())

In [106]:
#23
# Let's start with 2 topics
ldana = models.LdaModel(corpus=corpusna, num_topics=2, id2word=id2wordna, passes=10)
ldana.print_topics()

[(0,
  '0.011*"johnson" + 0.011*"churchill" + 0.010*"gandhi" + 0.010*"lincoln" + 0.009*"war" + 0.008*"boris" + 0.007*"indian" + 0.005*"independence" + 0.005*"british" + 0.005*"india"'),
 (1,
  '0.009*"mandela" + 0.008*"trump" + 0.008*"brandt" + 0.008*"tutu" + 0.007*"nelson" + 0.007*"washington" + 0.007*"south" + 0.006*"war" + 0.005*"german" + 0.005*"thatcher"')]

In [107]:
#24
# Let's try 3 topics
ldana = models.LdaModel(corpus=corpusna, num_topics=3, id2word=id2wordna, passes=10)
ldana.print_topics()

[(0,
  '0.012*"mandela" + 0.012*"gandhi" + 0.011*"trump" + 0.010*"brandt" + 0.009*"nelson" + 0.008*"indian" + 0.007*"south" + 0.006*"german" + 0.006*"india" + 0.005*"independence"'),
 (1,
  '0.023*"johnson" + 0.017*"boris" + 0.011*"thatcher" + 0.009*"party" + 0.009*"london" + 0.007*"conservative" + 0.006*"prime" + 0.006*"mrs" + 0.004*"leader" + 0.004*"policies"'),
 (2,
  '0.015*"war" + 0.013*"churchill" + 0.013*"lincoln" + 0.010*"tutu" + 0.010*"washington" + 0.006*"roosevelt" + 0.006*"south" + 0.005*"slavery" + 0.005*"desmond" + 0.004*"military"')]

In [108]:
#25
# Let's try 4 topics
ldana = models.LdaModel(corpus=corpusna, num_topics=4, id2word=id2wordna, passes=10)
ldana.print_topics()

[(0,
  '0.033*"churchill" + 0.021*"war" + 0.007*"winston" + 0.005*"last" + 0.005*"end" + 0.004*"british" + 0.004*"conservative" + 0.004*"military" + 0.004*"speech" + 0.004*"second"'),
 (1,
  '0.027*"mandela" + 0.020*"nelson" + 0.013*"thatcher" + 0.012*"south" + 0.008*"africa" + 0.007*"mrs" + 0.006*"apartheid" + 0.005*"economic" + 0.005*"policies" + 0.005*"conservative"'),
 (2,
  '0.012*"johnson" + 0.012*"gandhi" + 0.010*"trump" + 0.009*"tutu" + 0.008*"boris" + 0.008*"washington" + 0.007*"indian" + 0.006*"independence" + 0.006*"british" + 0.005*"london"'),
 (3,
  '0.019*"lincoln" + 0.016*"brandt" + 0.012*"war" + 0.009*"german" + 0.009*"germany" + 0.008*"roosevelt" + 0.007*"slavery" + 0.006*"union" + 0.005*"party" + 0.005*"civil"')]

## Identify Topics in Each Document

Out of the 9 topic models we looked at, the nouns and adjectives, 4 topic one made the most sense. So let's pull that down here and run it through some more iterations to get more fine-tuned topics.

In [109]:
#26
# Our final LDA model (for now)
ldana = models.LdaModel(corpus=corpusna, num_topics=4, id2word=id2wordna, passes=10)
ldana.print_topics()

[(0,
  '0.016*"mandela" + 0.014*"trump" + 0.013*"brandt" + 0.013*"tutu" + 0.012*"nelson" + 0.011*"south" + 0.008*"german" + 0.007*"africa" + 0.006*"desmond" + 0.006*"apartheid"'),
 (1,
  '0.019*"johnson" + 0.017*"lincoln" + 0.016*"gandhi" + 0.013*"boris" + 0.008*"independence" + 0.007*"london" + 0.006*"slavery" + 0.006*"indian" + 0.005*"party" + 0.005*"british"'),
 (2,
  '0.023*"churchill" + 0.017*"war" + 0.010*"thatcher" + 0.010*"roosevelt" + 0.006*"conservative" + 0.005*"winston" + 0.005*"economic" + 0.005*"mrs" + 0.005*"party" + 0.005*"fdr"'),
 (3,
  '0.019*"washington" + 0.009*"caste" + 0.009*"ambedkar" + 0.008*"indian" + 0.008*"constitution" + 0.007*"george" + 0.006*"india" + 0.006*"british" + 0.006*"untouchables" + 0.005*"independence"')]

These four topics look pretty decent. Let's settle on these for now.
* Topic 0: Political Leaders and Conflicts
* Topic 1: Desmond Tutu and South Africa
* Topic 2: Historical Figures and Social Issues
* Topic 3: Donald Trump and Economic Policies

In [110]:
#27
# Initialize an empty list to store the topic distributions for each document
topic_distributions = []

# Iterate over each document in the corpus and transform it to its topic distribution
for doc in corpusna:
    topic_distribution = ldana.get_document_topics(doc)
    topic_distributions.append(topic_distribution)

# Print the index of each document and its corresponding topic distribution
for index, topic_dist in zip(data_dtmna.index, topic_distributions):
    print(index, topic_dist)


abraham [(1, 0.9988266)]
ambedkar [(3, 0.9988647)]
boris [(1, 0.99893194)]
brandt [(0, 0.9989928)]
desmond [(0, 0.99855757)]
gandhi [(1, 0.9986254)]
mandela [(0, 0.9986791)]
margaret [(2, 0.99847627)]
roosevelt [(2, 0.998268)]
trump [(0, 0.998573)]
washington [(3, 0.99837375)]
winston [(2, 0.99880075)]


It seems like the transcripts are predominantly associated with specific topics based on the LDA model results:
* Abraham: Belongs strongly to Topic 1 with a probability of approximately 99.88%.
* Ambedkar: Belongs strongly to Topic 3 with a probability of approximately 99.89%.
* Boris: Belongs strongly to Topic 1 with a probability of approximately 99.89%.
* Brandt: Belongs strongly to Topic 0 with a probability of approximately 99.90%.
* Desmond: Belongs strongly to Topic 0 with a probability of approximately 99.86%.
* Gandhi: Belongs strongly to Topic 1 with a probability of approximately 99.86%.
* Mandela: Belongs strongly to Topic 0 with a probability of approximately 99.87%.
* Margaret: Belongs strongly to Topic 2 with a probability of approximately 99.85%.
* Roosevelt: Belongs strongly to Topic 2 with a probability of approximately 99.83%.
* Trump: Belongs strongly to Topic 0 with a probability of approximately 99.86%.
* Washington: Belongs strongly to Topic 3 with a probability of approximately 99.84%.
* Winston: Belongs strongly to Topic 2 with a probability of approximately 99.88%.

## Additional Exercises

1. Try further modifying the parameters of the topic models above and see if you can get better topics.


In [111]:
#28
# Modify the parameters of the LDA model
num_topics = 5  # Adjust the number of topics
passes = 30  # Adjust the number of passes
alpha = 'auto'  # Adjust the alpha parameter (e.g., 'auto', 'symmetric', or a list)
eta = 'auto'  # Adjust the eta parameter (e.g., 'auto', 'symmetric', or a scalar)
iterations = 400  # Adjust the number of iterations

# Train the LDA model with the modified parameters
lda_model = models.LdaModel(corpus=corpus, num_topics=num_topics, id2word=id2word,
                            passes=passes, alpha=alpha, eta=eta, iterations=iterations)

# Print the topics
lda_model.print_topics()

[(0,
  '0.019*"trump" + 0.010*"ambedkar" + 0.007*"caste" + 0.006*"indian" + 0.005*"donald" + 0.005*"india" + 0.004*"untouchables" + 0.004*"constitution" + 0.004*"business" + 0.003*"father"'),
 (1,
  '0.017*"churchill" + 0.017*"lincoln" + 0.014*"war" + 0.008*"roosevelt" + 0.007*"world" + 0.006*"slavery" + 0.005*"president" + 0.005*"great" + 0.004*"helped" + 0.004*"winston"'),
 (2,
  '0.016*"brandt" + 0.015*"gandhi" + 0.011*"washington" + 0.007*"german" + 0.007*"independence" + 0.006*"germany" + 0.006*"war" + 0.006*"british" + 0.005*"indian" + 0.005*"india"'),
 (3,
  '0.028*"tutu" + 0.014*"desmond" + 0.011*"south" + 0.006*"africa" + 0.006*"black" + 0.005*"apartheid" + 0.004*"peace" + 0.004*"critical" + 0.004*"africans" + 0.004*"justice"'),
 (4,
  '0.016*"mandela" + 0.015*"johnson" + 0.014*"boris" + 0.011*"nelson" + 0.010*"thatcher" + 0.006*"south" + 0.006*"london" + 0.006*"party" + 0.005*"mrs" + 0.005*"africa"')]

2. Create a new topic model that includes terms from a different [part of speech](https://www.ling.upenn.edu/courses/Fall_2003/ling001/penn_treebank_pos.html) and see if you can get better topics.

## Adverbs

In [128]:
#29
# Function to filter adverbs from text
def adverbs(text):
    '''Given a string of text, tokenize the text and pull out only the adverbs.'''
    is_adverb = lambda pos: pos[:2] == 'RB'
    tokenized = word_tokenize(text)
    all_adverbs = [word for (word, pos) in pos_tag(tokenized) if is_adverb(pos)]
    return ' '.join(all_adverbs)

In [113]:
#30
# Apply the nouns function to the transcripts to filter only on nouns
data_adverbs = pd.DataFrame(data_clean.transcript.apply(adverbs))
data_adverbs

transcript
abraham     neither well away hard also very often rather ...
ambedkar    also also shortly also not often very mainly a...
boris       also most increasingly never also often memora...
brandt      willy willy willy willy never later more incre...
desmond     frequently however together also increasingly ...
gandhi      shortly always always once also especially soo...
mandela     later jointly widely previously most not not n...
margaret    also october also thatcher rather however even...
roosevelt   often first very not later also not also not e...
trump       also donald donald then later then later manha...
washington  also often widely as well as especially poorly...
winston     best rarely reportedly also later officially l...

In [114]:
#31
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction import text
import pandas as pd  # Import pandas library

# Define additional stop words specific to this task
add_stop_words = ['like', 'im', 'know', 'just', 'dont', 'thats', 'right', 'people',
                  'youre', 'got', 'gonna', 'time', 'think', 'yeah', 'said']

# Combine additional stop words with the existing English stop words
stop_words = list(text.ENGLISH_STOP_WORDS.union(add_stop_words))

# Assuming data_adverbs contains the text data with adverbs
cv_adverbs = CountVectorizer(stop_words=stop_words, max_df=0.8)
data_cv_adverbs = cv_adverbs.fit_transform(data_adverbs.transcript)
data_dtm_adverbs = pd.DataFrame(data_cv_adverbs.toarray(), columns=cv_adverbs.get_feature_names_out())
data_dtm_adverbs.index = data_adverbs.index
data_dtm_adverbs


ably  abroad  abysmally  actively  actually  affectionately  \
abraham        0       0          0         0         0               0   
ambedkar       1       0          0         0         0               0   
boris          0       0          0         0         0               1   
brandt         0       1          0         0         1               0   
desmond        0       0          0         0         0               0   
gandhi         0       0          1         0         0               0   
mandela        0       0          0         1         0               0   
margaret       0       0          0         0         0               0   
roosevelt      0       0          0         0         0               0   
trump          0       1          0         0         0               0   
washington     0       1          0         0         0               0   
winston        0       0          0         0         0               0   

            afraid  ago  ahead  ambedkar  ...  unsuccessfully  usually  \
abraham          0    1      0         0  ...               0        0   
ambedkar         0    0      0         3  ...               0        0   
boris            0    0      0         0  ...               0        2   
brandt           0    0      0         0  ...               0        0   
desmond          0    0      0         0  ...               0        0   
gandhi           0    0      0         0  ...               0        0   
mandela          1    0      0         0  ...               0        0   
margaret         0    0      1         0  ...               0        0   
roosevelt        0    0      0         0  ...               0        0   
trump            0    0      0         0  ...               1        0   
washington       0    0      0         0  ...               0        0   
winston          0    0      0         0  ...               0        0   

            vigorously  visibly  warmly  whilst  wholeheartedly  widely  \
abraham              0        0       0       0               0       2   
ambedkar             0        0       0       0               0       1   
boris                0        0       1       0               0       0   
brandt               0        1       0       0               0       1   
desmond              0        0       0       0               0       0   
gandhi               1        0       0       1               0       0   
mandela              0        0       0       0               0       1   
margaret             0        0       0       0               0       1   
roosevelt            0        0       0       0               1       0   
trump                0        0       0       0               0       0   
washington           0        0       0       0               0       2   
winston              0        0       0       0               0       0   

            willy  yorktown  
abraham         0         0  
ambedkar        0         0  
boris           0         0  
brandt         14         0  
desmond         0         0  
gandhi          0         0  
mandela         0         0  
margaret        0         0  
roosevelt       0         0  
trump           0         0  
washington      0         1  
winston         0         0  

[12 rows x 136 columns]

In [115]:
#32
# Create the gensim corpus
corpus_adverbs = matutils.Sparse2Corpus(scipy.sparse.csr_matrix(data_dtm_adverbs.transpose()))

# Create the vocabulary dictionary
id2word_adverbs = dict((v, k) for k, v in cv_adverbs.vocabulary_.items())

In [116]:
#33
# Build the LDA model
# For 2 Topics
lda_adverbs = models.LdaModel(corpus=corpus_adverbs, num_topics=2, id2word=id2word_adverbs, passes=10)
lda_adverbs.print_topics()

[(0,
  '0.078*"willy" + 0.027*"widely" + 0.024*"shortly" + 0.024*"soon" + 0.019*"successfully" + 0.019*"ambedkar" + 0.019*"frequently" + 0.019*"nearly" + 0.018*"politically" + 0.018*"abroad"'),
 (1,
  '0.040*"increasingly" + 0.030*"eventually" + 0.027*"thatcher" + 0.026*"highly" + 0.022*"far" + 0.022*"einstein" + 0.021*"especially" + 0.021*"forward" + 0.021*"newton" + 0.019*"widely"')]

In [117]:
#34
# Build the LDA model
# For 3 Topics
lda_adverbs = models.LdaModel(corpus=corpus_adverbs, num_topics=3, id2word=id2word_adverbs, passes=10)
lda_adverbs.print_topics()

[(0,
  '0.054*"thatcher" + 0.044*"eventually" + 0.034*"away" + 0.033*"widely" + 0.033*"lincoln" + 0.024*"relatively" + 0.024*"initially" + 0.024*"strongly" + 0.023*"merely" + 0.014*"largely"'),
 (1,
  '0.050*"increasingly" + 0.036*"forward" + 0.035*"previously" + 0.020*"largely" + 0.020*"highly" + 0.020*"eventually" + 0.020*"einstein" + 0.020*"widely" + 0.020*"newton" + 0.020*"successfully"'),
 (2,
  '0.064*"willy" + 0.028*"frequently" + 0.028*"increasingly" + 0.024*"successfully" + 0.024*"especially" + 0.019*"widely" + 0.019*"soon" + 0.019*"shortly" + 0.019*"newton" + 0.019*"far"')]

In [118]:
#35
# Build the LDA model
# For 4 Topics
lda_adverbs = models.LdaModel(corpus=corpus_adverbs, num_topics=4, id2word=id2word_adverbs, passes=10)
lda_adverbs.print_topics()

[(0,
  '0.044*"thatcher" + 0.035*"frequently" + 0.035*"increasingly" + 0.035*"far" + 0.027*"eventually" + 0.024*"especially" + 0.019*"away" + 0.019*"instead" + 0.019*"largely" + 0.019*"donald"'),
 (1,
  '0.042*"newton" + 0.033*"soon" + 0.033*"increasingly" + 0.033*"widely" + 0.032*"einstein" + 0.032*"successfully" + 0.031*"rarely" + 0.026*"especially" + 0.023*"forward" + 0.023*"highly"'),
 (2,
  '0.035*"widely" + 0.035*"lincoln" + 0.035*"ambedkar" + 0.024*"nearly" + 0.024*"increasingly" + 0.024*"politically" + 0.024*"eventually" + 0.024*"strongly" + 0.024*"forever" + 0.024*"shortly"'),
 (3,
  '0.180*"willy" + 0.029*"successfully" + 0.016*"abroad" + 0.016*"widely" + 0.016*"immediately" + 0.016*"skillfully" + 0.016*"shortly" + 0.016*"partly" + 0.016*"firstly" + 0.016*"fluently"')]

In [119]:
# Final LDA Model for Adverbs
lda_adverbs = models.LdaModel(corpus=corpus_adverbs, num_topics=4, id2word=id2word_adverbs, passes=80)
lda_adverbs.print_topics()

[(0,
  '0.152*"willy" + 0.035*"far" + 0.035*"increasingly" + 0.035*"especially" + 0.024*"successfully" + 0.024*"frequently" + 0.013*"abroad" + 0.013*"immediately" + 0.013*"largely" + 0.013*"firstly"'),
 (1,
  '0.052*"eventually" + 0.052*"thatcher" + 0.032*"widely" + 0.032*"away" + 0.032*"lincoln" + 0.022*"relatively" + 0.022*"strongly" + 0.022*"initially" + 0.022*"unexpectedly" + 0.022*"highly"'),
 (2,
  '0.051*"increasingly" + 0.027*"frequently" + 0.027*"ambedkar" + 0.027*"previously" + 0.018*"successfully" + 0.018*"forward" + 0.018*"instead" + 0.018*"shortly" + 0.018*"partly" + 0.018*"usually"'),
 (3,
  '0.045*"soon" + 0.031*"newton" + 0.031*"rarely" + 0.031*"widely" + 0.031*"especially" + 0.017*"einstein" + 0.017*"abroad" + 0.017*"skillfully" + 0.017*"successfully" + 0.017*"deeply"')]

* Topic 0: Adverbs indicating frequency or intensity (e.g., especially, rarely, highly).
* Topic 1: Adverbs indicating time or immediacy (e.g., soon, shortly, eventually).
* Topic 2: Adverbs related to degree or extent (e.g., increasingly, far, largely).
* Topic 3: A mix of adverbs along with some proper nouns (e.g., willy, Einstein, Lincoln), possibly indicating a less coherent topic.

## Noun Adjective and Verb

In [120]:
def nouns_adj_verbs(text):
    '''Given a string of text, tokenize the text and pull out only the nouns, adjectives, and verbs.'''
    is_noun_adj_verb = lambda pos: pos[:2] == 'NN' or pos[:2] == 'JJ' or pos[:2] == 'VB'
    tokenized = word_tokenize(text)
    nouns_adj_verbs = [word for (word, pos) in pos_tag(tokenized) if is_noun_adj_verb(pos)]
    return ' '.join(nouns_adj_verbs)


In [121]:
# Apply the new function to the transcripts to filter only on nouns, adjectives, and verbs
data_nouns_adj_verbs = pd.DataFrame(data_clean.transcript.apply(nouns_adj_verbs))

# Create a new document-term matrix using nouns, adjectives, and verbs
cvnav = CountVectorizer(stop_words=stop_words)
data_cvnav = cvnav.fit_transform(data_nouns_adj_verbs.transcript)
data_dtmnav = pd.DataFrame(data_cvnav.toarray(), columns=cvnav.get_feature_names_out())
data_dtmnav.index = data_nouns_adj_verbs.index

data_dtmnav


abandoning  abe  abhorred  ability  abitur  able  abolish  \
abraham              0    1         1        1       0     1        0   
ambedkar             0    0         0        0       0     0        0   
boris                0    0         0        0       0     0        0   
brandt               0    0         0        0       1     2        0   
desmond              0    0         0        0       0     1        0   
gandhi               0    0         0        0       0     0        0   
mandela              0    0         0        1       0     1        0   
margaret             1    0         0        0       0     0        0   
roosevelt            0    0         0        0       0     2        1   
trump                0    0         0        0       0     1        0   
washington           0    0         0        0       0     2        0   
winston              0    0         0        0       0     1        0   

            abolitionists  abomination  abraham  ...  yorker  young  \
abraham                 1            0       11  ...       0      1   
ambedkar                0            0        0  ...       0      2   
boris                   0            0        0  ...       0      1   
brandt                  0            0        0  ...       0      2   
desmond                 0            1        0  ...       1      0   
gandhi                  0            0        0  ...       0      1   
mandela                 0            0        0  ...       0      1   
margaret                0            0        0  ...       0      0   
roosevelt               0            0        1  ...       0      0   
trump                   0            0        1  ...       0      0   
washington              0            0        0  ...       0      1   
winston                 0            0        0  ...       0      1   

            youngster  yousafzai  youth  zeal  zealand  zelnickova  zone  zulu  
abraham             0          0      0     0        0           0     0     0  
ambedkar            0          0      0     0        0           0     0     0  
boris               0          0      0     0        0           0     0     0  
brandt              0          0      1     1        0           0     0     0  
desmond             0          0      0     0        0           0     0     0  
gandhi              1          0      0     0        0           0     0     1  
mandela             1          1      0     0        1           0     0     0  
margaret            0          0      0     0        0           0     1     0  
roosevelt           0          0      0     0        0           0     0     0  
trump               0          0      0     0        0           1     0     0  
washington          0          0      0     0        0           0     0     0  
winston             0          0      0     0        0           0     0     0  

[12 rows x 3506 columns]

In [122]:
# Create the gensim corpus
corpusnav = matutils.Sparse2Corpus(scipy.sparse.csr_matrix(data_dtmnav.transpose()))

# Create the vocabulary dictionary
id2wordnav = dict((v, k) for k, v in cvnav.vocabulary_.items())

In [123]:
# Let's start with 2 topics
ldanav = models.LdaModel(corpus=corpusnav, num_topics=2, id2word=id2wordnav, passes=10)
ldanav.print_topics()

[(0,
  '0.009*"churchill" + 0.008*"lincoln" + 0.008*"war" + 0.007*"trump" + 0.006*"washington" + 0.005*"world" + 0.005*"president" + 0.004*"roosevelt" + 0.004*"great" + 0.004*"famous"'),
 (1,
  '0.008*"gandhi" + 0.007*"brandt" + 0.007*"mandela" + 0.007*"johnson" + 0.006*"boris" + 0.006*"south" + 0.006*"tutu" + 0.006*"nelson" + 0.004*"indian" + 0.004*"africa"')]

In [124]:
# Let's start with 3 topics
ldanav = models.LdaModel(corpus=corpusnav, num_topics=3, id2word=id2wordnav, passes=10)
ldanav.print_topics()

[(0,
  '0.011*"churchill" + 0.010*"brandt" + 0.009*"gandhi" + 0.009*"trump" + 0.008*"war" + 0.005*"german" + 0.005*"famous" + 0.004*"world" + 0.004*"thatcher" + 0.004*"party"'),
 (1,
  '0.015*"lincoln" + 0.012*"tutu" + 0.011*"washington" + 0.006*"south" + 0.006*"desmond" + 0.006*"slavery" + 0.005*"war" + 0.004*"george" + 0.004*"president" + 0.004*"new"'),
 (2,
  '0.011*"johnson" + 0.011*"mandela" + 0.010*"boris" + 0.008*"nelson" + 0.006*"roosevelt" + 0.005*"london" + 0.005*"ambedkar" + 0.005*"south" + 0.004*"world" + 0.004*"political"')]

In [125]:
# Let's start with 4 topics
ldanav = models.LdaModel(corpus=corpusnav, num_topics=4, id2word=id2wordnav, passes=10)
ldanav.print_topics()

[(0,
  '0.016*"gandhi" + 0.015*"mandela" + 0.011*"nelson" + 0.009*"indian" + 0.008*"south" + 0.008*"india" + 0.007*"ambedkar" + 0.006*"independence" + 0.006*"africa" + 0.006*"caste"'),
 (1,
  '0.017*"churchill" + 0.012*"washington" + 0.010*"war" + 0.007*"thatcher" + 0.006*"mrs" + 0.005*"world" + 0.005*"famous" + 0.005*"british" + 0.005*"winston" + 0.004*"conservative"'),
 (2,
  '0.019*"johnson" + 0.017*"trump" + 0.016*"boris" + 0.007*"london" + 0.005*"party" + 0.004*"campaign" + 0.004*"donald" + 0.004*"vote" + 0.003*"political" + 0.003*"oxford"'),
 (3,
  '0.013*"lincoln" + 0.012*"brandt" + 0.010*"tutu" + 0.009*"war" + 0.006*"roosevelt" + 0.006*"german" + 0.006*"germany" + 0.005*"world" + 0.005*"south" + 0.005*"desmond"')]

In [126]:
# Our final LDA model (for now)
ldanav = models.LdaModel(corpus=corpusnav, num_topics=3, id2word=id2wordnav, passes=30)
ldanav.print_topics()

[(0,
  '0.013*"lincoln" + 0.011*"trump" + 0.006*"roosevelt" + 0.006*"thatcher" + 0.005*"president" + 0.005*"war" + 0.005*"great" + 0.004*"slavery" + 0.004*"world" + 0.004*"mrs"'),
 (1,
  '0.009*"brandt" + 0.009*"mandela" + 0.009*"johnson" + 0.008*"boris" + 0.008*"tutu" + 0.007*"nelson" + 0.007*"south" + 0.004*"german" + 0.004*"party" + 0.004*"london"'),
 (2,
  '0.016*"churchill" + 0.014*"gandhi" + 0.010*"washington" + 0.010*"war" + 0.007*"independence" + 0.007*"british" + 0.005*"indian" + 0.004*"india" + 0.004*"winston" + 0.004*"george"')]

In [127]:
# Let's take a look at which topics each transcript contains
corpus_transformed = ldanav[corpusna]
topics_and_indices = []
for doc_topics in corpus_transformed:
    topic_index, _ = max(doc_topics, key=lambda x: x[1])
    topics_and_indices.append(topic_index)

# Now, zip the topics with the document indices
result = list(zip(topics_and_indices, data_dtmna.index))

# Print the result
print(result)

[(1, 'abraham'), (1, 'ambedkar'), (1, 'boris'), (1, 'brandt'), (1, 'desmond'), (1, 'gandhi'), (1, 'mandela'), (1, 'margaret'), (1, 'roosevelt'), (1, 'trump'), (1, 'washington'), (0, 'winston')]


* Topic 0: Highlights contemporary political leaders such as Boris Johnson, Donald Trump, and Margaret Thatcher
* Topic 1: emphasizes influential figures in social and political movements, including Mahatma Gandhi, Nelson Mandela, and Desmond Tutu
* Topic 2: Delves into historical figures and events, featuring Winston Churchill, Abraham Lincoln, and Willy Brandt